### PR TITLE
PUBDEV-8497: grouping variable 

### PR DESCRIPTION
**Pull request by koniecsveta**
https://h2oai.atlassian.net/browse/PUBDEV-8497

Option "grouping_variable" to specify a grouping variable to group the data and provide separate sets of plots by grouping variable values.

for example:

py:
path_5

R:
Rplot

### CODE REVIEW

File: h2o-py/tests/testdir_misc/pyunit_ice_plot.py [hunk 1]
```diff
@@ -1,9 +1,8 @@
 from __future__ import print_function
 
-import os
 import sys
 
-sys.path.insert(1, os.path.join("..", "..", ".."))
+sys.path.insert(1,"../../")
```
Comment (johnygomez) [inline h2o-py/tests/testdir_misc/pyunit_ice_plot.py:5]: Is this change really needed? This may break the functionality on various OS

File: h2o-py/h2o/explanation/_explain.py [hunk 2]
```diff
@@ -1230,6 +1230,49 @@ def pd_multi_plot(
 
 def _center(col):
     col[:] = col - col[0]
+def _prepare_grouping_frames(frame, grouping_variable):
+    if (grouping_variable not in frame.names):
```
Comment (johnygomez) [inline h2o-py/h2o/explanation/_explain.py:1234]: nit/style: It's stylistically preferred to not wrap conditions with parenthesis except for long expressions. Ref e.g. https://google.github.io/styleguide/pyguide.html#33-parentheses

File: h2o-py/h2o/explanation/_explain.py [hunk 3]
```diff
@@ -1230,6 +1230,49 @@ def pd_multi_plot(
...
+        raise ValueError("Grouping variable '" + grouping_variable + "' is not present in frame!")
+    if (not frame[grouping_variable].isfactor()[0]):
+        raise ValueError("Grouping variable has to be categorical!")
+    categories = frame[grouping_variable].categories()
+    if len(categories) > 10:
+        raise ValueError("Grouping variable option is supported only for variables with 10 or fewer levels!")
```
Comment (johnygomez) [inline h2o-py/h2o/explanation/_explain.py:1240]: `10` looks like a magic number, is there reasoning for this? Should it be also configurable somehow?

File: h2o-py/h2o/explanation/_explain.py [hunk 4]
```diff
@@ -1230,6 +1230,49 @@ def pd_multi_plot(
...
+    frames = list()
+    for curr_category in categories:
+        key = "tmp_" + curr_category + str(random.randint(0,1000))
+        expr = "(tmp= " + key + "(rows " + frame.frame_id + " (==(cols " + frame.frame_id + " [" + str(frame.names.index(grouping_variable)) + "] ) '" + curr_category + "') ))"
```
Comment (johnygomez) [inline h2o-py/h2o/explanation/_explain.py:1244]: Oh no, what is this? 🙂  Is it going to be evaluated in python? Also I think much more readable alternative is to do string interpolation using `string.format()`

File: h2o-py/h2o/explanation/_explain.py [hunk 5]
```diff
@@ -1230,6 +1230,49 @@ def pd_multi_plot(
...
+        h2o.rapids(expr)
+        frames.append(h2o.get_frame(key))
+    return frames
+
+def _handle_grouping(frame, grouping_variable, save_plot_path, model, column, target, max_levels, figsize, colormap):
+    frames = _prepare_grouping_frames(frame, grouping_variable)
+    result = list()
+    if save_plot_path is not None:
+        len_path = len(save_plot_path)
```
Comment (johnygomez) [inline h2o-py/h2o/explanation/_explain.py:1253]: nit: If the goal is to extract an extension of the filename, it seems like a more robust alternative is to use [`os.path.splitext`](https://docs.python.org/3.8/library/os.path.html#os.path.splitext), which does exactly that.
I.e.
```python
root_path, ext = os.path.splitext(save_plot_path)
```

File: h2o-py/h2o/explanation/_explain.py [hunk 6]
```diff
@@ -1230,6 +1230,49 @@ def pd_multi_plot(
...
+    i = 0
+    for curr_frame in frames:
```
Comment (johnygomez) [inline h2o-py/h2o/explanation/_explain.py:1255]: nit/optional: It's fine to use explicit counter, but a little more pythonic would be to use `enumerate` this way
```python
for idx, curr_frame in enumerate(frames):
    ...
```
and you can spare the counter

**Review by johnygomez — COMMENTED** [review]
I have few nits, mostly of stylistic character. I however don't fully understand how the frame for a specific group is created. Otherwise, it seems the functionality is correct :+1:

File: h2o-py/h2o/explanation/_explain.py (hunk 2 above)
Comment (koniecsveta) [reply h2o-py/h2o/explanation/_explain.py]: fixed

File: h2o-py/h2o/explanation/_explain.py (hunk 3 above)
Comment (koniecsveta) [reply h2o-py/h2o/explanation/_explain.py]: it's a requirement specification, it shouldn't be configurable.

File: h2o-py/h2o/explanation/_explain.py (hunk 4 above)
Comment (koniecsveta) [reply h2o-py/h2o/explanation/_explain.py]: nono, this is a rapids expression to be evaluated on java backend in this way: if there is a categorical variable for example color with values blue, black and white -> it will create 3 separate frames, one per each category: first with data only with blue rows, second with data with only black rows and third with data with only white rows. I don't mind changing to `string.format()`

File: h2o-py/h2o/explanation/_explain.py (hunk 5 above)
Comment (koniecsveta) [reply h2o-py/h2o/explanation/_explain.py]: thanks!

File: h2o-py/h2o/explanation/_explain.py [hunk 7]
```diff
@@ -1302,6 +1347,9 @@ def ice_plot(
     if frame.type(column) == "string":
         raise ValueError("String columns are not supported!")
 
+    if grouping_variable != None:
```
Comment (johnygomez) [inline h2o-py/h2o/explanation/_explain.py:1350]: nit: Use `is not` when comparing to `None` rather than equality sign. TL;DR It's possible that object equals `None` even though it's not really `None`. The `is` operator confirms that the variable is actually an instance of `None`, which is always just single object.

File: h2o-py/h2o/explanation/_explain.py [hunk 8]
```diff
@@ -1380,10 +1428,12 @@ def ice_plot(
                          label="Partial Dependence")
 
         _add_histogram(frame, column)
-        plt.title("Individual Conditional Expectation for \"{}\"\non column \"{}\"{}".format(
+        group_label = kwargs.get('group_label')
```
Comment (johnygomez) [inline h2o-py/h2o/explanation/_explain.py:1431]: ultranit: You can provide default value, and can avoid ternary later.
```suggestion
        group_label = kwargs.get("group_label", "")
```

**Review by johnygomez — APPROVED** [review]
LGTM. Thanks for changes 🙂

File: h2o-r/h2o-package/R/explain.R [hunk 9]
```diff
@@ -2393,7 +2435,10 @@ h2o.ice_plot <- function(model,
                          max_levels = 30,
                          show_pdp = TRUE,
                          binary_response_scale = c("response", "logodds"),
-                         centered = FALSE) {
+
+                         centered = FALSE,
+                         grouping_variable=NULL
```
Comment (tomasfryda) [inline h2o-r/h2o-package/R/explain.R:2440]:
```suggestion
                         grouping_variable = NULL
```

**Review by tomasfryda — APPROVED** [review]
Looks good, thank you @koniecsveta !

**Comment by ledell** [comment]
The standard way in H2O to make a column arg would be `something_column` so I wonder if it should be `grouping_column` or `group_column`?  Though I like the sound of "grouping variable" since I think that's what you might call it outside of H2O.

**Review by ledell — APPROVED** [review]

**Comment by ledell** [comment]
Can you make the description of `grouping_variable` more clear in that it should be the name of a column in the data?
